### ISSUE TITLE
large file compression examples

### ISSUE THREAD

**Issue by zixin530** [borisbrodski/sevenzipjbinding#15]
Can you provide some detailed code snippets about large file compressions please? I tried several methods(for instance implementing my own ISequentialInStream) and it gives me jvm crash and core dump error. Thank you! Really appreciate it.

**Comment by zixin530** [comment]
Thank you for your reply. 
 I tried your example and it worked well with a small file(200kb). For a large file like 20 mb, it's reading very slow and I don't know how long it will take(I've waited 25+ mins). So I tried to use a buffer array. It works great with single file.  It looks like the following:
 
`int read(byte[] data) {
      int bufferSize = 1024;
      byte[] buffer = new byte[bufferSize];
      try {
          int result = inputStream.read(buffer);
      } catch (IOException ee) {
      }
      for(int i = 0; i < buffer.length; i++ ){
           data[i] = buffer[i];       // copy buffer to data array
      }
      if(result == -1)
             return 0; //marked as end of stream
      return buffer.length
}
`
 **The problem** I encountered is that when there are multiple files(large-20mb or small-50kb), it gives me jvm crash and IINStream class not found error. So I can only compress one file but not multiple files. I dont know why this happens. 
![jvm_crash](https://user-images.githubusercontent.com/22278256/32594243-e08a178e-c4f0-11e7-949f-1e637c5cd72f.png)
The only time it works it when the files are all very small(about 100 bytes). I attached my implementation in a text file. Thank you again. 
[MyImplementation.txt](https://github.com/borisbrodski/sevenzipjbinding/files/1456763/MyImplementation.txt)

**Comment by borisbrodski** [comment]
You still reading the entire file into the memory

```java
    byte[] content = Files.readAllBytes(fl.get(index).toPath());
```

You can get the size of the file without reading it content. See java.io.File.

In `read(byte[] data)` I would simply call

```java
    result = in.read(data);
    if (result <= 0) {
        return 0;
    }
    return result;  // <- Actual count of bytes
```

It is important to return the actual count of bytes, that was read. `return buffer.length` is wrong.

Also you can't assume, that the size of the `data` array will be always greater than 1024.

**Comment by zixin530** [comment]
Sorry I'm not following it. are you saying that I read the entire file into memory because I did item.setDataSize((long)content)?  I change the "content" to fl.get(index).length. I also change the read(byte[] data) based on your suggestion, but it still doesn't work for multiple files... Thank you so much for your patience.

**Comment by borisbrodski** [comment]
No, you do `byte[] content = Files.readAllBytes(fl.get(index).toPath());`. This reads the entire file into the memory.

I have a lot of tests, that create huge archives with hundreds of items. So, it should also with for you.

I more suggestion: try to remove all directory-items from your list. Compress only files. You can still use paths like "dir1/dir2/file.txt" to create your directory structure. But don't try to compress items, that represent directory explicitly.

**Comment by zixin530** [comment]
So I used  `item.setDataSize(fl.get(index).length());` to get the size of the file. And after some tests, i found that multiple large files can be compressed if I set the solid size to a small value(1000 or so).     
 Like this: `outArchive.setSolidSize(1024);`  If I dont do this line or set the solid size to a large number like 2e9, it gives me jvm crash. I do need the solid compression to be on for my project. Is it because the heap size is not big enough? I've increased heap size in intellij and tomcat(I checked the log info. it's confirmed), but it doesn't work. Thank you. 
![123](https://user-images.githubusercontent.com/22278256/32647155-ed0066ea-c62a-11e7-9fab-0ed6b38c9cb6.png)

**Comment by borisbrodski** [comment]
Ah, now I understand. So the compression works for small data sizes. That's good.

Have you tried to compress your files with the 7-zip with such large solid block size. It well may be, that for 2e9 solid block 7-zip required much more memory, that just 2e9. I would run 7-zip (in the same version, that you are using with 7-Zip-JBinding) in monitor the memory consumption.

**Comment by zixin530** [comment]
I used 7zip GUI on windows with my 20ish-mb files, and I set it to ultra compression and lzma method. For the block size, I choose the option "solid compression".It tells me that it consumes 2g ram.

**Comment by borisbrodski** [comment]
Actually, if you set "solid block" to true you don't have to set the block size, since the entire archive will be stored in a single block. May be, if you set both "solid block" to true and the block size to some value, 7-zip compression engine get confused. Try to just set "solid block" to true and see, what you get.

Check, if you get a solid block archive or not.

**Comment by zixin530** [comment]
So I tried set the "solid block" to be true and comment out the solid size line. However, it doesn't give me a solid bock archive. It is just a regular 7z archive. Here is the screenshot. 
 
![image](https://user-images.githubusercontent.com/22278256/32880019-1c6a013c-cae7-11e7-8029-373721bfe382.png)

I also try to comment out the `outArchive.setSolidFiles(1);` , but it gives me JVM crash.
Thank you

**Comment by wshidaijingshen** [comment]
> So I used `item.setDataSize(fl.get(index).length());` to get the size of the file. And after some tests, i found that multiple large files can be compressed if I set the solid size to a small value(1000 or so).
> Like this: `outArchive.setSolidSize(1024);` If I dont do this line or set the solid size to a large number like 2e9, it gives me jvm crash. I do need the solid compression to be on for my project. Is it because the heap size is not big enough? I've increased heap size in intellij and tomcat(I checked the log info. it's confirmed), but it doesn't work. Thank you.
> ![123](https://user-images.githubusercontent.com/22278256/32647155-ed0066ea-c62a-11e7-9fab-0ed6b38c9cb6.png)

You did right  that we should both  "setSolid(true)" AND "setSolidSize(8192)"  is the only way to avoid the  weird JVM crash []!
Very lucky I found this answer in your report,and it drives me crazy  to try to solve this problem for all day.
 
I have to to say:  what you mean 2e9 is too big? 2e9 = 512, and it's very small size.  We can set it as  8192  or bigger . 8192 is fine for me and it works well . Here is  what I do :

`if (outArchive instanceof IOutFeatureSetSolid) {
            ((IOutFeatureSetSolid) outArchive).setSolid(solid);
            ((IOutFeatureSetSolid) outArchive).setSolidSize(8192L);
        }

/*and we should override the getStream()  and return just use like  new  RandomAccessFileInStream() , no need to implements ourselves . borisbrodski  told me this secret in another issue I opened,hhh..
*/

@Override
            public ISequentialInStream getStream(int index) {
                if (files.get(index).isDirectory()) {
                    return null;
                }

                // very big file ,such as 2G，must use this class RandomAccessFileInStream，or the  OOM will occurs，see at  https://github.com/borisbrodski/sevenzipjbinding/issues/32

                RandomAccessFile raf = new RandomAccessFile(files.get(index), "rw");
                return new RandomAccessFileInStream(raf);
            }
`

Also,with tomcat container, its memory is limited so this problem occurs when someone  code in a web container like tomcat.But it never occurs  when outside of tomcat container ,such as  in a plain java  program.

It confuses me that borisbrodski (the author of the git project?) didn't  give the accurate answer.
But still thanks to you for developing this wonderful  project and share it with us all.
Thank U,guys.
God bless US
